evalml/utils/woodwork_utils.py

```diff
@@ -27,6 +28,32 @@ def _list_to_pandas(list):
     return _numpy_to_pandas(np.array(list))
 
 
+def _new_infer_boolean_nullable_func(series):
+    if pdtypes.is_bool_dtype(series.dtype) and not pdtypes.is_categorical_dtype(
+        series.dtype,
+    ):
+        return True
+    elif pdtypes.is_object_dtype(series.dtype):
+        series_no_null = series.dropna()
+        try:
+            series_no_null_unq = set(series_no_null)
+            if series_no_null_unq in [
+                {False, True},
+                {True},
+                {False},
+            ]:
+                return True
+        except TypeError:  # Necessary to check for non-hashable values because of object dtype consideration
+            return False
+    return False
+
+
+ww.type_system.update_inference_function(
+    "BooleanNullable",
+    inference_function=_new_infer_boolean_nullable_func,
+)
```

Comment:
This is wacky. If I understand correctly, we're directly changing woodwork inference behavior during init of the evalml package? Why is this necessary for evalml? It seems odd to fork off of standard woodwork behavior, it may cause headaches in the future.

Reply:
This was the reason that the `update_inference_function` was added in Woodwork, so that inference functions could be changed/added whenever necessary.

Unfortunately the alternative is to not only update every test that runs into this Boolean inference issue, but also to update several components like the `LabelEncoder` that run into reverse mapping issues. In order for the new Boolean inference to be supported here, there needs to be an underlying change in cases where previously `Categorical` variables will be treated as `Boolean` instead.

Luckily any downstream `Boolean` transformations have to go through this inference first, so this can act as a sort of "gatekeeper" for this behaviour, for now.

Reply:
@ParthivNaresh is what EvalML considers "boolean" different from what WW does now? and what is the difference? Just curious why inference shouldn't be standard between WW and EvalML.